### PR TITLE
Time delay estimation

### PR DESCRIPTION
**Pull request by Sosnowsky**

### CODE REVIEW

**Comment by Sosnowsky** [comment]
Major refactor.
@auroradh please check that the code works fine in some more shots and that you are comfortable with the new interface (ask if there are any questions). I checked for the reference acmod shot.

File: fppanalysis/two_dim_velocity_estimates.py [hunk 1]
```diff
@@ -96,15 +114,16 @@ def __init__(self, ds, estimation_options: EstimationOptions):
...
-
-        from pathos.multiprocessing import ProcessPool
+        self.tde_delegator = tde.TDEDelegator(
+            estimation_options.method,
+            estimation_options.get_time_delay_options(),
+            estimation_options.cache,
+        )
+        self.pixels = [[PixelData() for _ in range_r] for _ in range_z]
 
-        pool = ProcessPool(estimation_options.num_cores)
```
Comment (gregordecristoforo) [inline fppanalysis/two_dim_velocity_estimates.py:103]: Ok, so we completely remove the multiprocessing functionality?

File: guides/how_to_use_2D_velocity_estimate.ipynb [hunk 2]
```diff
@@ -1,4 +1,494 @@
 {
```
Comment (gregordecristoforo) [inline guides/how_to_use_2D_velocity_estimate.ipynb:1]: I tried to open the notebook separately and got the following error: 
Invalid Notebook
The Notebook Does Not Appear to Be Valid JSON
Using nbformat v5.8.0 and nbconvert v7.3.1

Does anyone else have the same issue?

File: guides/how_to_use_2D_velocity_estimate.ipynb [hunk 3]
```diff
@@ -1,4 +1,494 @@
...
+<<<<<<< HEAD
```
Comment (gregordecristoforo) [inline guides/how_to_use_2D_velocity_estimate.ipynb:2]: I suppose this line has to be removed

File: guides/how_to_use_2D_velocity_estimate.ipynb [hunk 4]
```diff
@@ -1,4 +1,494 @@
...
+        "ax.set_ylabel(\"Z [cm]\")\n",
+        "ax.set_xlim([min(R_coordinates[:, 0]) - 0.5, max(R_coordinates[:, 0]) + 0.5])\n",
+        "ax.set_ylim(min(Z_coordinates[0]) - 0.5, max(Z_coordinates[0] + 0.5))\n",
+        "ax.set_title(\n",
+        "    f\"{min_threshold} \"\n",
+        "    r\"$ < \\tilde{\\mathrm{I}} < $ \"\n",
+        "    f\"{max_threshold}\"\n",
+        ")\n",
+        "plt.show()"
+=======
```
Comment (gregordecristoforo) [inline guides/how_to_use_2D_velocity_estimate.ipynb:491]: same here

File: guides/how_to_use_2D_velocity_estimate.ipynb [hunk 5]
```diff
@@ -917,11 +1407,34 @@
...
       "text/plain": [
        "<Figure size 640x480 with 2 Axes>"
+>>>>>>> main
```
Comment (gregordecristoforo) [inline guides/how_to_use_2D_velocity_estimate.ipynb:1410]: and here

File: guides/how_to_use_2D_velocity_estimate.ipynb [hunk 6]
```diff
@@ -917,11 +1407,34 @@
...
       ]
-     },
-     "metadata": {},
-     "output_type": "display_data"
     }
+<<<<<<< HEAD
```
Comment (gregordecristoforo) [inline guides/how_to_use_2D_velocity_estimate.ipynb:1413]: and here

File: guides/how_to_use_2D_velocity_estimate.ipynb [hunk 7]
```diff
@@ -917,11 +1407,34 @@
...
+      "name": "python",
+      "nbconvert_exporter": "python",
+      "pygments_lexer": "ipython3",
+      "version": "3.10.12"
+    },
+    "colab": {
+      "provenance": [],
+      "toc_visible": true
+    }
+=======
```
Comment (gregordecristoforo) [inline guides/how_to_use_2D_velocity_estimate.ipynb:1437]: and here

File: guides/how_to_use_2D_velocity_estimate.ipynb [hunk 8]
```diff
@@ -963,20 +1476,8 @@
    "display_name": "Python 3 (ipykernel)",
    "language": "python",
    "name": "python3"
+>>>>>>> main
```
Comment (gregordecristoforo) [inline guides/how_to_use_2D_velocity_estimate.ipynb:1479]: and here

File: guides/how_to_use_2D_velocity_estimate.ipynb [hunk 9]
```diff
@@ -963,20 +1476,8 @@
...
-   "pygments_lexer": "ipython3",
-   "version": "3.10.12"
-  }
- },
- "nbformat": 4,
- "nbformat_minor": 2
-}
+  "nbformat": 4,
+  "nbformat_minor": 0
+}
```
Comment (gregordecristoforo) [inline guides/how_to_use_2D_velocity_estimate.ipynb:1483]: But I really appreciate having a notebook as a guide :)

**Review by gregordecristoforo — CHANGES_REQUESTED** [review]
Really nice work! Just a few questions and minor comments on my side. I would however strongly recommend that someone else looks at this PR who is going to work with it frequently in the future (@auroradh, @sajidah-ahmed ?)

File: fppanalysis/time_delay_estimation.py [hunk 10]
```diff
@@ -198,40 +204,394 @@ def get_error(params):
     return avg, minimization.x
 
 
-def estimate_time_delay_ccmax(
-    x: np.ndarray, y: np.ndarray, dt: float, interpolate: bool = False
+class TDEOptions(ABC):
```
[the remark this answers is not in the document]
Comment (Sosnowsky) [reply fppanalysis/time_delay_estimation.py:207]: Fixed

File: fppanalysis/two_dim_velocity_estimates.py (hunk 1 above)
Comment (Sosnowsky) [reply fppanalysis/two_dim_velocity_estimates.py:103]: Yep

File: guides/how_to_use_2D_velocity_estimate.ipynb (hunk 2 above)
Comment (Sosnowsky) [reply guides/how_to_use_2D_velocity_estimate.ipynb]: Fixed

**Review by gregordecristoforo — APPROVED** [review]

**Pull request merged**